File: Stripe/STPPushProvisioningDetails.m
//
//  STPPushProvisioningDetails.m
//  Stripe
//
//  Created by Jack Flintermann on 9/26/18.
//  Copyright © 2018 Stripe, Inc. All rights reserved.
//

#import "STPPushProvisioningDetails.h"

#import "NSDictionary+Stripe.h"

@interface STPPushProvisioningDetails ()
    
@property (nonatomic, readwrite) NSString *cardId;
@property (nonatomic, readwrite) BOOL livemode;
@property (nonatomic, readwrite) NSData *encryptedPassData;
@property (nonatomic, readwrite) NSData *activationData;
@property (nonatomic, readwrite) NSData *ephemeralPublicKey;
@property (nonatomic, readwrite, copy) NSDictionary *allResponseFields;

@end

@implementation STPPushProvisioningDetails

+ (instancetype)detailsWithCardId:(NSString *)cardId
                         livemode:(BOOL)livemode
                encryptedPassData:(NSData *)encryptedPassData
                   activationData:(NSData *)activationData
               ephemeralPublicKey:(NSData *)ephemeralPublicKey {
    STPPushProvisioningDetails *details = [[self alloc] init];
    details.cardId = cardId;
    details.livemode = livemode;
    details.encryptedPassData = encryptedPassData;
    details.activationData = activationData;
    details.ephemeralPublicKey = ephemeralPublicKey;
    return details;
}
    
#pragma mark  - STPAPIResponseDecodable
    
+ (instancetype)decodedObjectFromAPIResponse:(NSDictionary *)response {
    NSDictionary *dict = [response stp_dictionaryByRemovingNulls];
    if (!dict) {
        return nil;
    }
    
    // required fields
    NSString *cardId = [dict stp_stringForKey:@"card"];
    BOOL livemode = [dict stp_boolForKey:@"livemode" or:NO];
    NSString *encryptedPassString = [dict stp_stringForKey:@"contents"];
    NSData *encryptedPassData = encryptedPassString ? [[NSData alloc] initWithBase64EncodedString:encryptedPassString options:0] : nil;
    
    NSString *activationString = [dict stp_stringForKey:@"activation_data"];
    NSData *activationData = activationString ? [[NSData alloc] initWithBase64EncodedString:activationString options:0] : nil;
    
    NSString *ephemeralPublicKeyString = [dict stp_stringForKey:@"ephemeral_public_key"];
    NSData *ephemeralPublicKeyData = ephemeralPublicKeyString ? [[NSData alloc] initWithBase64EncodedString:ephemeralPublicKeyString options:0] : nil;
    
    if (cardId == nil || encryptedPassData == nil || activationData == nil || ephemeralPublicKeyData == nil) {
        return nil;
    }
    
    STPPushProvisioningDetails *details = [self detailsWithCardId:cardId
                                                         livemode:livemode
                                                encryptedPassData:encryptedPassData
                                                   activationData:activationData
                                               ephemeralPublicKey:ephemeralPublicKeyData];
    details.allResponseFields = dict;
    
    return details;
}

    
#pragma mark - Equality

- (BOOL)isEqual:(STPPushProvisioningDetails *)details {
    return [self isEqualToDetails:details];
}

- (NSUInteger)hash {
    return [self.activationData hash];
}

- (BOOL)isEqualToDetails:(STPPushProvisioningDetails *)details {
    if (self == details) {
        return YES;
    }
    
    if (!details || ![details isKindOfClass:self.class]) {
        return NO;
    }
    
    return [self.activationData isEqualToData:details.activationData];
}

@end

File: Stripe/STPPushProvisioningDetails.swift
//
//  STPPushProvisioningDetails.swift
//  Stripe
//
//  Created by Jack Flintermann on 9/26/18
//  Copyright © 2018 Stripe, Inc. All rights reserved.
//

import Foundation

class STPPushProvisioningDetails: NSObject, STPAPIResponseDecodable {
    let cardId: String
    let livemode: Bool
    let encryptedPassData: Data
    let activationData: Data
    let ephemeralPublicKey: Data

    required init(
        cardId: String,
        livemode: Bool,
        encryptedPass encryptedPassData: Data,
        activationData: Data,
        ephemeralPublicKey: Data
    ) {
        self.cardId = cardId
        self.livemode = livemode
        self.encryptedPassData = encryptedPassData
        self.activationData = activationData
        self.ephemeralPublicKey = ephemeralPublicKey
        super.init()
    }
    private(set) var allResponseFields: [AnyHashable: Any] = [:]

    // MARK: - STPAPIResponseDecodable
    class func decodedObject(fromAPIResponse response: [AnyHashable: Any]?) -> Self? {
        guard
            let dict = (response as NSDictionary?)?.stp_dictionaryByRemovingNulls() as NSDictionary?
        else {
            return nil
        }

        // required fields
        let cardId = dict.stp_string(forKey: "card")
        let livemode = dict.stp_bool(forKey: "livemode", or: false)
        let encryptedPassString = dict.stp_string(forKey: "contents")
        let encryptedPassData =
            encryptedPassString != nil
            ? Data(base64Encoded: encryptedPassString ?? "", options: []) : nil

        let activationString = dict.stp_string(forKey: "activation_data")
        let activationData =
            activationString != nil ? Data(base64Encoded: activationString ?? "", options: []) : nil

        let ephemeralPublicKeyString = dict.stp_string(forKey: "ephemeral_public_key")
        let ephemeralPublicKeyData =
            ephemeralPublicKeyString != nil
            ? Data(base64Encoded: ephemeralPublicKeyString ?? "", options: []) : nil

        if cardId == nil || encryptedPassData == nil || activationData == nil
            || ephemeralPublicKeyData == nil
        {
            return nil
        }

        if let encryptedPassData = encryptedPassData, let activationData = activationData,
            let ephemeralPublicKeyData = ephemeralPublicKeyData
        {
            let details = self.init(
                cardId: cardId ?? "",
                livemode: livemode,
                encryptedPass: encryptedPassData,
                activationData: activationData,
                ephemeralPublicKey: ephemeralPublicKeyData)
            details.allResponseFields = dict as! [AnyHashable: Any]
            return details
        }
        return nil
    }

    // MARK: - Equality
    override func isEqual(_ object: Any?) -> Bool {
        if let details = object as? STPPushProvisioningDetails {
            return isEqual(to: details)
        }
        return false
    }

    override var hash: Int {
        return activationData.hashValue
    }

    func isEqual(to details: STPPushProvisioningDetails) -> Bool {
        return details.activationData == self.activationData
    }
}
